Support for Active Record 5

myabc opened the pull request:
- You probably want to merge this into the `activerecord-5.0` branch rather than `master`. PR is here for discussion/tracking (unlike https://github.com/SchemaPlus/schema_plus_core/pull/14 though, it might not be so difficult to support AR 4.2 + 5.0 in the same project).

myabc committed:
Update .gemspec to depend on schema_plus_core 2.0

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Only test against ActiveRecord 5.0, Ruby 2.3.0

myabc committed:
Update deprecated #tables to #data_sources

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Explicitly override PostgreSQLAdapter#remove_index

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Specify Migration version for CommandRecorder spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

ronen commented:
@myabc Thanks for working on this!  Since you marked it as Work In Progress I'll ignore it for now until you say it's ready.  

BTW I presume you're using the soon-to-be-released schema_plus_core 2.0.0 that @boazy created in SchemaPlus/schema_plus_core/pull/14 ?

myabc commented:
@ronen I have specs passing locally for MySQL; PostgreSQL. SQLite3 still needs some work, hence the PR title (It's probably the RDBMS I'm least familiar with).

> BTW I presume you're using the soon-to-be-released schema_plus_core 2.0.0 that @boazy created in SchemaPlus/schema_plus_core#14 ?

Having this in my `Gemfile.local` works fine:

``` ruby
gem 'schema_plus_core', github: 'boazy/schema_plus_core',
                        branch: 'activerecord-5.0'
```

I'm just trying to get the same working on Travis.

myabc commented on spec/command_recorder_spec.rb:
```diff
@@ -37,7 +37,7 @@ def before(env)
 
     it "does not fail when reverting" do
-      migration = Class.new ::ActiveRecord::Migration do
+      migration = Class.new ::ActiveRecord::Migration[5.0] do
```
this allows the spec to pass, but is probably not the right solution:

`env.options[:index] = { name: 'whatever' } if enabled_middleware(AddIndexBefore, env)`

~~(in this case, the call to `enabled_middleware` always returns false)~~

I need to better understand what the desired behaviour should be.

ronen replied:
Strange that `enabled_middleware` always returns false.  Are you sure?  (Actually it should return either a class or `nil`.  See `spec/support/enableable.rb`

The intent of the spec is to make sure that the middleware doesn't cause a `:down` migration to fail; I probably introduced this test to fix a bug.  In particular, it should exercise the code in `lib/schema_plus/indexes/active_record/migration/command_recorder.rb`

myabc replied:
> Strange that enabled_middleware always returns false. Are you sure? (Actually it should return either a class or nil. See spec/support/enableable.rb

Sorry, I had meant falsey (i.e. `nil`), not `false`.

> (in this case, the call to enabled_middleware always returns false)

This statement was actually incorrect. I took a fresh look at it this morning. All appending `[5.0]` to `::ActiveRecord::Migration` does is fix a deprecation warning.

myabc commented:
@ronen Do you have a standard way of telling Travis to use a different upstream branch. I had considered configuring `.travis.yml` to write out a `Gemfile.local, but I know it's auto-generated, thus would be overriden on the first matrix run.

ronen commented:
@myabc 

> Do you have a standard way of telling Travis to use a different upstream branch.

The thing you did with Gemfile.local ought to work I think... but in any case it's moot since I've just released schema_plus_core 2.0.0

myabc committed:
Explicitly override SQLite3Adapter#remove_index

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add sanity check to Schema::SQLite3::Indexes

Currently the `Result` returned includes auto-generated indexes
e.g. `sqlite_autoindex_ar_internal_metadata_1` for which the `sql`
value will always be `nil`.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Revert "Update deprecated #tables to #data_sources"

In these cases, `#tables` is actually the appropriate method.

This reverts commit e543c18af17cf381fb96e92a3fdb19cdf2186d38.

myabc commented on lib/schema_plus/indexes/middleware/schema.rb:
```diff
@@ -8,7 +8,7 @@ def after(env)
 
             env.connection.exec_query("SELECT name, sql FROM sqlite_master WHERE type = 'index'").map do |row|
-              if (desc_columns = row['sql'].scan(/['"`]?(\w+)['"`]? DESC\b/).flatten).any?
+              if row['sql'] && (desc_columns = row['sql'].scan(/['"`]?(\w+)['"`]? DESC\b/).flatten).any?
```
@ronen I'm not completely sure why the query returns auto-generated indexes here.

ronen replied:
I'm OK with adding the guard if the tests show it's needed.

myabc commented:
@ronen I think this is now ready for an initial review.

ronen commented on .travis.yml:
```diff
@@ -5,11 +5,11 @@
 sudo: false
 rvm:
-- 2.1.5
+- 2.3.0
```
May as well use 2.3.1 , since that's the current officially stable release.

ronen commented on gemfiles/activerecord-5.0/Gemfile.base:
```diff
@@ -1,3 +1,3 @@
 eval File.read File.expand_path('../../Gemfile.base', __FILE__)
 
-gem "activerecord", "~> 4.2.6"
+gem "activerecord", ">= 5.0.0.beta1", "< 5.1"
```
why not "~> 5.0" ?

myabc replied:
@ronen I think this is an issue with `schema_dev`. This project's `Gemfile` specifies `~> 5.0`.

ronen replied:
oh duh of course it's generated by `schema_dev`.  Now that AR 5.0 is out of beta i've switched that to `~> 5.0.0` in `schema_dev` and pushed a new release

ronen commented on lib/schema_plus/indexes/active_record/connection_adapters/sqlite3_adapter.rb:
```diff
@@ -0,0 +1,19 @@
+
+        # Extends rails' remove_index to include this options:
+        #   :if_exists
+        def remove_index(table_name, *args)
```
Seems not very DRY to have this same code repeated in abstract_adapter, postgresql_adapter and sqlite3_adapter.   I think it would be good to create something like `lib/schema_plus/indexes/remove_if_exists.rb` that defines a module `SchemaPlus::Indexes::RemoveIfExists` containing the `remove_index` method.  Then each of the adapters could just do

```
module SchemaPlus::Indexes
  module ActiveRecord
    module ConnectionAdapters
      module Sqlite3Adapter
         include RemoveIfExists
      end
    end
  end
end
```

Though if that doesn't work for some annoying metaprogramming reason, could be slightly less DRY (though arguably clearer) and instead of including the module, define a method `SchemaPlus::Indexes::RemoveIndex.if_exists` which could be used in each adapter as

```
def remove_index(table_name, *args)
   super table_name, *args if SchemaPlus::Indexes::RemoveIndex.if_exists(table_name, args)
end
```

Also, if the method in `abstract_adapter` is only applicable to mysql since the others need to override it explicitly, then maybe move it to mysql to minimize confusion?

What do you think about all that?

ronen commented on schema_dev.yml:
```diff
@@ -1,7 +1,7 @@
 ruby:
-    - 2.1.5
+    - 2.3.0
```
Oh, _this_ is where the 2.3.1 would be specified :)

ronen commented:
@myabc thanks again for working on this!

Duplicating my own comment in https://github.com/SchemaPlus/schema_plus_foreign_keys/pull/11#issuecomment-240022048 ...

My only other overall question is, as AR keeps increasing its support for indexes, what (if any) functionality of schema_plus_indexes is now natively supported by AR? Though potentially with different syntax?

(Revealing my ignorance about AR 5, I don't know what has been added. Part of why I've been so behind the ball on the AR 5 upgrades to SchemaPlus is that I'm not currently using rails much in my day jobs.)

myabc committed:
Bump tested Ruby to 2.3.1

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
DRY up: Extract #remove_index override into Module

Signed-off-by: Alex Coles <alex@alexbcoles.com>

boazy commented:
Is this PR ready to be merged?
Once it gets released, schema_plus_columns could be released as well.
Next comes schema_validations, but seems to take a little bit more work.

ronen commented:
Sorry for the delay.  @myabc thanks for the update!

I've merged and adjusted the dependencies because this actually still works for AR 4.2 and 5.0.  Since there are no new features or breaking changes, this is just a patch release :)  

I've released v0.2.4
